File: frontend-v2/src/features/chat/lib/compaction-view.test.ts
import { describe, expect, it } from "vitest"
import type { MessageWithParts } from "@/shared/types/api"
import { buildCompactionViews } from "./compaction-view"
import { buildAssistantContentView } from "./content-view"
import { mergeTurns } from "./turn-view"

function reply(id: string, extra: Partial<MessageWithParts> = {}): MessageWithParts {
  return { id, role: "assistant", session_id: "s", created_at: "", parts: [], ...extra }
}

const request = (auto: boolean): MessageWithParts => reply("compact", {
  role: "user", agent: "compaction", parts: [
    { type: "text", id: "empty", text: "", synthetic: true },
    { type: "compaction", id: "marker", auto },
  ],
})
const summary = (extra: Partial<MessageWithParts> = {}): MessageWithParts => reply("summary", {
  agent: "compaction", parent_id: "compact", summary: false,
  parts: [{ type: "text", id: "summary-text", text: "## Goal\nKeep the important constraints." }],
  ...extra,
})
const answer = reply("answer", {
  finish: "stop", tokens: { input: 20, output: 5, total: 25, cache: 0, limit: 1000, cost: 0, context: 25 },
  parts: [{ type: "text", id: "answer-text", text: "The task is complete.", channel: "final" }],
})

describe("compaction is a process item throughout its lifecycle", () => {
  it("shows a running item as soon as the internal request arrives", () => {
    expect(buildCompactionViews([request(true)], true)).toEqual([
      { id: "compact", status: "running", summary: "" },
    ])
  })

  it("folds streaming text before summary=true is persisted, without creating answer prose", () => {
    const messages = [request(true), summary()]
    expect(buildCompactionViews(messages, true)[0]).toMatchObject({ id: "compact", status: "running" })
    const content = buildAssistantContentView(messages, true)
    expect(content.hasFinal).toBe(false)
    expect(content.progress).toEqual([])
    expect(content.workEvents).toEqual([])
  })

  it("keeps one stable item when completion and the next loop step arrive", () => {
    const messages = [request(true), summary({ summary: true, finish: "stop" }), answer]
    expect(buildCompactionViews(messages, true)).toHaveLength(1)
    expect(buildCompactionViews(messages, true)[0].status).toBe("completed")
    const content = buildAssistantContentView(messages, false)
    expect(content.finalMessageId).toBe("answer")
    expect(content.finalText).toBe("The task is complete.")
    expect(content.progress).toEqual([])
  })

  it("recovers a summary whose request is outside the loaded page", () => {
    const completed = summary({ summary: true, finish: "stop" })
    expect(buildCompactionViews([completed], false)).toEqual(
      buildCompactionViews([request(true), completed], false),
    )
  })

  it("supports old stored summaries and completed descriptors without text", () => {
    expect(buildCompactionViews([summary({ agent: undefined, summary: true, finish: "stop" })], false)[0].status).toBe("completed")
    const committed = request(true)
    committed.parts = [{ type: "compaction", id: "marker", replacement_id: "replacement" }]
    expect(buildCompactionViews([committed], false)[0]).toMatchObject({ status: "completed", summary: "" })
  })

  it("does not call a failed or interrupted partial summary a successful answer", () => {
    expect(buildCompactionViews([request(true), summary({ error: { message: "provider failed" } })], false)[0].status).toBe("failed")
    expect(buildCompactionViews([request(true), summary()], false)[0].status).toBe("interrupted")
    expect(buildAssistantContentView([summary({ error: { message: "provider failed" } })], false).hasFinal).toBe(false)
  })

  it("does not revive an unfinished old attempt when a new loop step starts", () => {
    expect(buildCompactionViews([request(true), summary(), reply("next")], true)[0].status).toBe("interrupted")
  })

  it("keeps separate retries visible without duplicating a request and its summary", () => {
    const second = { ...request(true), id: "compact-2" }
    const views = buildCompactionViews([
      request(true), summary({ summary: true, finish: "stop" }), second,
      summary({ id: "summary-2", parent_id: "compact-2" }),
    ], true)
    expect(views.map((view) => view.status)).toEqual(["completed", "running"])
  })
})

describe("turn grouping around context optimization", () => {
  it("keeps automatic compaction and continuation inside the same visible turn", () => {
    const before = reply("before", { finish: "tool_calls" })
    const continueMessage = reply("continue", {
      role: "user", parts: [{ type: "text", id: "continue-text", text: "Continue", synthetic: true }],
    })
    const turns = mergeTurns([before, request(true), summary({ finish: "stop" }), continueMessage, answer])
    expect(turns).toHaveLength(1)
    expect(turns[0].kind).toBe("assistant")
    if (turns[0].kind !== "assistant") throw new Error("Expected assistant turn")
    expect(turns[0].meta.messageId).toBe("answer")
    expect(turns[0].messages.map((message) => message.id)).toEqual(["before", "compact", "summary", "answer"])
  })

  it("does not replace reply metadata with summary tokens or failure", () => {
    const turns = mergeTurns([answer, request(true), summary({ error: { message: "failed" } })])
    if (turns[0].kind !== "assistant") throw new Error("Expected assistant turn")
    expect(turns[0].meta.messageId).toBe("answer")
    expect(turns[0].meta.tokens).toEqual(answer.tokens)
    expect(turns[0].meta.error).toBeUndefined()
  })

  it("keeps manual optimization with the previous turn's process and answer", () => {
    const turns = mergeTurns([answer, request(false), summary({ summary: true, finish: "stop" })])
    expect(turns.map((turn) => turn.kind)).toEqual(["assistant"])
    if (turns[0].kind !== "assistant") throw new Error("Expected assistant turn")
    expect(buildAssistantContentView(turns[0].messages, false).finalMessageId).toBe("answer")
    expect(turns[0].meta.messageId).toBe("answer")
  })
})

File: mobile/test/features/chat/compaction_view_test.dart
import 'package:bossip_mobile/features/chat/utils/compaction_view.dart';
import 'package:bossip_mobile/features/chat/utils/content_view.dart';
import 'package:bossip_mobile/features/chat/utils/turn_view.dart';
import 'package:bossip_mobile/shared/models/message.dart';
import 'package:bossip_mobile/shared/models/message_part.dart';
import 'package:flutter_test/flutter_test.dart';

import 'compaction_fixtures.dart';

void main() {
  test('streaming summary is process before summary flag is set', () {
    final messages = [compactRequest(), compactSummary()];
    final row = buildChatRows(messages).single as AssistantTurnData;
    expect(row.lastReply, isNull);
    expect(row.bodyText, isEmpty);
    expect(row.notices, isEmpty);
    final content = buildAssistantContentView(messages, true);
    expect(content.hasFinal, isFalse);
    expect(content.workEvents, isEmpty);
    expect(
      buildCompactionViews(messages, true).single.status,
      CompactionStatus.running,
    );
  });

  test('manual and auto compression preserve the answer and reply actions', () {
    for (final auto in [true, false]) {
      final original = answer();
      final messages = [
        original,
        compactRequest(auto: auto),
        compactSummary(finish: 'stop'),
      ];
      final row = buildChatRows(messages).single as AssistantTurnData;
      expect(row.lastReply, same(original));
      expect(row.lastMessageId, original.id);
      expect(row.finish, 'stop');
      final content = buildAssistantContentView(messages, false);
      expect(content.finalText, 'Answer');
      expect(content.finalMessageId, original.id);
      expect(content.workEvents, isEmpty);
      expect(
        buildCompactionViews(messages, false).single.status,
        CompactionStatus.completed,
      );
    }
  });

  test('continued loop keeps multiple optimizations in one turn', () {
    final messages = [
      answer(text: 'Working', finish: 'tool_calls'),
      compactRequest(),
      compactSummary(finish: 'stop'),
      answer(id: 'm04', text: 'Continue', finish: 'tool_calls'),
      compactRequest(id: 'm05'),
      compactSummary(id: 'm06', parent: 'm05', finish: 'stop'),
      answer(id: 'm07', text: 'Final answer'),
    ];
    expect(buildChatRows(messages), hasLength(1));
    expect(buildCompactionViews(messages, false).map((v) => v.id), [
      'm02',
      'm05',
    ]);
    final content = buildAssistantContentView(messages, false);
    expect(content.finalText, 'Final answer');
    expect(content.progress.map((v) => v.text), ['Working', 'Continue']);
  });

  test(
    'failure and interruption never replace the answer or mark its turn failed',
    () {
      for (final finish in ['error', 'aborted']) {
        final messages = [
          answer(),
          compactRequest(),
          compactSummary(finish: finish),
        ];
        final row = buildChatRows(messages).single as AssistantTurnData;
        expect(row.error, isNull);
        expect(buildAssistantContentView(messages, false).finalText, 'Answer');
        expect(
          buildCompactionViews(messages, false).single.status,
          finish == 'error'
              ? CompactionStatus.failed
              : CompactionStatus.interrupted,
        );
      }
    },
  );

  test('history pages recognize lone summaries, requests and old rows', () {
    final legacy = compactSummary(finish: 'stop', legacy: true);
    expect(buildAssistantContentView([legacy], false).hasFinal, isFalse);
    expect(buildCompactionViews([legacy], false).single.id, 'm02');
    final request = compactRequest(replacement: 'replacement-1');
    expect(
      buildCompactionViews([request], false).single.status,
      CompactionStatus.completed,
    );
    expect(
      buildCompactionViews([compactRequest()], false).single.status,
      CompactionStatus.interrupted,
    );
    expect(buildCompactionViews([], true), isEmpty);
  });

  test(
    'more than 200 messages keep real user boundaries and the next answer',
    () {
      final history = [
        for (var i = 0; i < 125; i++) ...[
          ChatMessage(
            id: 'u$i',
            sessionId: 's1',
            role: 'user',
            parts: [TextPart(id: 'p$i', text: 'Task $i')],
          ),
          answer(id: 'a$i'),
        ],
        compactRequest(),
        compactSummary(finish: 'stop'),
        const ChatMessage(
          id: 'next-user',
          sessionId: 's1',
          role: 'user',
          parts: [TextPart(id: 'next-text', text: 'Next task')],
        ),
        answer(id: 'next-answer', text: 'Continue normally'),
      ];
      final rows = buildChatRows(history);
      expect(rows.whereType<UserRowData>(), hasLength(126));
      expect(rows.whereType<AssistantTurnData>(), hasLength(126));
      expect((rows.last as AssistantTurnData).bodyText, 'Continue normally');
    },
  );
}
